pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -940,12 +941,55 @@ public PinotResourceManagerResponse deleteBrokerTenantFor(String tenantName) {
   /**
    * Schema APIs
    */
-  public void addOrUpdateSchema(Schema schema) {
+  public void addSchema(Schema schema, boolean shouldOverride) {
+    ZNRecord record = SchemaUtils.toZNRecord(schema);
+    String schemaName = schema.getSchemaName();
+    Schema oldSchema = ZKMetadataProvider.getSchema(_propertyStore, schemaName);
+
+    if (oldSchema != null && !shouldOverride) {
+      throw new RuntimeException("Schema exists. Not overriding it as requested.");
+    }
+
+    if (schema.equals(oldSchema)) {
+      LOGGER.info("New schema is the same with the existing schema. Not updating.");
+      return;
+    }
+
+    PinotHelixPropertyStoreZnRecordProvider propertyStoreHelper =
+        PinotHelixPropertyStoreZnRecordProvider.forSchema(_propertyStore);
+    propertyStoreHelper.set(schemaName, record);
+  }
+
+  public void updateSchema(Schema schema, boolean shouldAutoReload)
+      throws TableNotFoundException {
     ZNRecord record = SchemaUtils.toZNRecord(schema);
-    String name = schema.getSchemaName();
+    String schemaName = schema.getSchemaName();
+    Schema oldSchema = ZKMetadataProvider.getSchema(_propertyStore, schemaName);
+
+    if (oldSchema == null) {
+      throw new RuntimeException("Schema did not exist.");
+    }
+
+    if (schema.equals(oldSchema)) {
+      LOGGER.info("New schema is the same with the existing schema. Not updating.");
+      return;
+    }
+
     PinotHelixPropertyStoreZnRecordProvider propertyStoreHelper =
         PinotHelixPropertyStoreZnRecordProvider.forSchema(_propertyStore);
-    propertyStoreHelper.set(name, record);
+    propertyStoreHelper.set(schemaName, record);
+
+    if (oldSchema != null && shouldAutoReload && schema.isBackwardCompatibleWith(oldSchema)) {
+      // Automatically reload all tables that have this schema
+      LOGGER.info("Reloading all tables with schema name: {}", schemaName);
+      for (String tableName : getAllTables()) {
+        if (schemaName.equals(getTableSchema(tableName).getSchemaName())) {
```

Comment:
We actually have some legacy tables with schema names that are different from table names. We could do a cleanup, but it may happen for other people since it's technically allowed.

Reply:
We log warnings for such tables for a long time. We need to figure out a way to enforce table name to be the same as schema, and enforce OFFLINE tables to have a schema. Multiple features rely on this.
We have thousands of tables in the same cluster, and this can cause quite big performance issue...

Reply:
Agree some enforcement should be in place on the stuff you mentioned. One slight optimization is, after retrieving schemaZNRecord, we just return schema name by ZNRecord id, instead of parsing the json and get schema name from the constructed schema. Or, we just reload by schema name, and declare we do not support auto reload for those "invalid" tables..

Reply:
Calling `getTableSchema()` will fetch the table config if schema cannot be found, which is not acceptable for big cluster.
We can just reload the tables with raw name the same as schema name. For the "invalid" tables, auto-reload won't have any effect.